app/coffee/EventLogger.coffee

```diff
@@ -0,0 +1,46 @@
+logger = require 'logger-sharelatex'
+
+# keep track of message counters to detect duplicate and out of order events
+# messsage ids have the format "UNIQUEHOSTKEY-COUNTER"
+
+EVENT_LOG_COUNTER = {}
+EVENT_LOG_TIMESTAMP = {}
+EVENT_COUNT = 0
+
+module.exports = EventLogger =
+
+	MAX_EVENTS_BEFORE_CLEAN: 100000
+	MAX_STALE_TIME_IN_MS: 3600 * 1000
+
+	checkEventOrder: (message_id, message) ->
+		return if typeof(message_id) isnt 'string'
+		[key, count] = message_id.split("-", 2)
+		count = parseInt(count, 10)
+		if !count # ignore checks if counter is not present
+			return
+		# store the last count in a hash for each host
+		previous = EventLogger._storeEventCount(key, count)
+		if !previous? || count == (previous + 1)
+			return # order is ok
+		if (count == previous)
```

Comment:
I think this could be `<=`, since it's possible it's duplicate, and out of order?

Reply:
Actually, I see the else branch covers that, so ignore.